3/99_hw/fast.go

```diff
@@ -0,0 +1,166 @@
+
+func scanBrowsers(browser []byte, seenBrowsers *[]string, uniqueBrowsers *int, user *User) {
+	//fmt.Printf("Checking browser: %s\n", browser)
+	
+	isAndroid := bytes.Contains(browser, androidBrowser)
+	isMSIE := bytes.Contains(browser, msieBrowser)
+	if isAndroid || isMSIE {
+		//fmt.Printf("Browser found: android=%v, MSIE=%v\n", isAndroid, isMSIE)
+		user.IsAndroid = user.IsAndroid || isAndroid
+		user.IsMSIE = user.IsMSIE || isMSIE
+
+		browserStr := string(browser)
+		if !slices.Contains(*seenBrowsers, browserStr) {
+			//fmt.Printf("FAST New browser: %s\n", browser)
+			*seenBrowsers = append(*seenBrowsers, browserStr)
+			*uniqueBrowsers++
+		}
+	}
+}
+
+func scanUser(userStr *[]byte, seenBrowsers *[]string, uniqueBrowsers *int, user *User) error {
+	user.IsAndroid = false
+	user.IsMSIE = false
+
+	var token *bytes.Buffer
+	var tokenProcessing int 
+	var arrayProcessing int
+	var browserProcessing, nameProcessing, emailProcessing int
+
+	for _, c := range *userStr {
+
+		switch c {			
+			case '"':
+				switch tokenProcessing {
+				case 0:
+					tokenProcessing++
+					token = dataPool.Get().(*bytes.Buffer)
+				case 1:
+					//fmt.Printf("token: %s\n", token)
+					if nameProcessing == 1 {
+						result := token.String()
+						nameProcessing++
+						user.Name = result
+						nameProcessing++
+						//fmt.Printf("--> User name: %v\n", user.Name)
+					} else if emailProcessing == 1 {
+						result := token.String()
+						emailProcessing++
+						user.Email = strings.Replace(result, "@", " [at] ", 1)
+						emailProcessing++
+						//fmt.Printf("--> User email: %v\n", user.Email)
+					} else if browserProcessing == 1 && arrayProcessing > 0 {
+						scanBrowsers(token.Bytes(), seenBrowsers, uniqueBrowsers, user)
+					} else {
+						result := token.String()
+						switch result {
+						case "name":
+							nameProcessing++
+						case "email":
+							emailProcessing++
+						case "browsers":
+							browserProcessing++
+						}
+					}
+
+					tokenProcessing = 0
+					token.Reset()
+					dataPool.Put(token)
+				}
+			case '[':
+				if tokenProcessing == 0 {
+					arrayProcessing++
+				}
+			case ']':
+				if tokenProcessing == 0 {
+					arrayProcessing = 0
+				}
+			default:
+				if tokenProcessing == 1 {
+					token.WriteByte(c)
+				}
+		}
+	}
+
+	if len(user.Name) == 0 || len(user.Email) == 0 {
+		return fmt.Errorf("no data found")
+	}
+
+	return nil
+}
+
+// вам надо написать более быструю оптимальную этой функции
+func FastSearch(out io.Writer) {
+	/*
+		!!! !!! !!!
+		обратите внимание - в задании обязательно нужен отчет
+		делать его лучше в самом начале, когда вы видите уже узкие места, но еще не оптимизировалм их
+		так же обратите внимание на команду в параметром -http
+		перечитайте еще раз задание
+		!!! !!! !!!
+	*/
+	
+	file, err := os.Open(filePath)
+	if err != nil {
+		panic(err)
+	}
+
+	uniqueBrowsers := 0
+	seenBrowsers := make([]string, 0, 200)
+	foundUsers := make([]string, 0, 200)
+	user := User{}
```

Comment:
Пустые, неинициализированные структуры принято объявлять так
```suggestion
	var user User
```